crates/admin/src/cluster_controller/replicated_loglet_controller.rs

```diff
@@ -0,0 +1,656 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use enumset::EnumSet;
+use rand::seq::IteratorRandom;
+use rand::RngCore;
+use tracing::info;
+
+use restate_core::metadata_store::ReadError;
+use restate_core::network::{NetworkSender, TransportConnect};
+use restate_types::cluster_controller::{SchedulingPlan, TargetLogletState};
+use restate_types::logs::metadata::{LogletParams, SegmentIndex};
+use restate_types::logs::LogId;
+use restate_types::nodes_config::{
+    LogServerConfig, NodeConfig, NodesConfiguration, Role, StorageState,
+};
+use restate_types::replicated_loglet::{
+    NodeSet, ReplicatedLogletId, ReplicatedLogletParams, ReplicationProperty,
+};
+use restate_types::{GenerationalNodeId, PlainNodeId, Versioned};
+
+#[derive(Debug, thiserror::Error)]
+#[error("failed reading scheduling plan from metadata store: {0}")]
+pub struct BuildError(#[from] ReadError);
+
+#[derive(Debug, Clone)]
+pub struct LogletControllerConfig {
+    replication: ReplicationProperty,
+}
+
+pub struct ReplicatedLogletController {
+    config: LogletControllerConfig,
+    scheduling_plan: SchedulingPlan,
+    observed_cluster_state: ObservedClusterState,
+    // we can add observed partition state, to better inform leader selection
+
+    // task_center: TaskCenter,
+    // metadata_store_client: MetadataStoreClient,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum LogletEffect {
+    AddLogletSegment,
+    SealLogletSegment,
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+struct ObservedClusterState {
+    /// Logs that need to be provisioned. We don't care about partition mapping at the moment.
+    logs: BTreeSet<LogId>,
+    /// Worker nodes that are considered healthy candidates for sequencer placement.
+    healthy_workers: BTreeMap<PlainNodeId, GenerationalNodeId>,
+    /// Log servers that are considered healthy and self-report as read-writable, candidates for loglet replication.
+    healthy_log_servers: BTreeMap<PlainNodeId, GenerationalNodeId>,
+    /// Any node that was previously known but isn't seen as healthy.
+    dead_nodes: BTreeSet<PlainNodeId>,
+
+    // todo: can we replace both sets of servers above with a simple?
+    // nodes: NodesConfiguration,
+
+    // Extra info - is it necessary for scheduling decisions?
+    node_roles: BTreeMap<PlainNodeId, EnumSet<Role>>,
+    log_server_states: BTreeMap<PlainNodeId, StorageState>,
+    // outstanding_actions: [create loglet, seal loglet, etc. to avoid duplication if they take more than one iteration to complete]
+}
+
+/// The loglet controller is responsible for safely configuring loglet segments based on overall
+/// policy and the available log servers, and for transitioning segments to new node sets as cluster
+/// members come and go. It repeatedly decides on a loglet scheduling plan which it writes to the
```

Comment:
I guess with loglet scheduling plan you are referring to a `LogletConfig` with the respective `LogletParams`? Or do you have something different in mind?